yt/data_objects/data_containers.py

```diff
@@ -462,25 +470,24 @@ def write_out(self, filename, fields=None, format="%0.16e"):
         if self._key_fields is None:
             raise ValueError
 
-        field_order = self._key_fields
+        field_order = [("index", k) for k in self._key_fields]
         diff_fields = [field for field in fields if field not in field_order]
         field_order += diff_fields
         field_order = sorted(self._determine_fields(field_order))
-        field_types = {u for u, v in field_order}
-
-        if len(field_types) != 1:
-            diff_fields = self._determine_fields(diff_fields)
-            req_ftype = self._determine_fields(self._key_fields[0])[0][0]
-            f_type = {f for f in diff_fields if f[0] != req_ftype}
-            msg = (
-                "Field type %s of the supplied field %s is inconsistent"
-                " with field type '%s'."
-                % ([f[0] for f in f_type], [f[1] for f in f_type], req_ftype)
-            )
-            raise YTException(msg)
 
+        field_shapes = {}
         for field in field_order:
-            self[field]
+            field_shapes[field] = self[field].shape
+
+        # Check all fields have the same shape
+        all_field_shapes = {shape for shape in field_shapes.values()}
+        if len(all_field_shapes) != 1:
+            err_msg = ["Got fields with different number of elements:\n"]
+            for shape in all_field_shapes:
+                these_fields = [field for field, val in field_shapes.items()]
```

Comment:
It seems the error message isn't quite correct, try for example:
```python
ds = yt.testing.fake_random_ds(16, particles=10)
sp = ds.all_data()
sp.write_out("/tmp/foo.txt", fields=[("all", "particle_ones")])
```
which returns
```
YTException: Got fields with different number of elements:

         [('all', 'particle_ones'), ('index', 'dx'), ('index', 'dy'), ('index', 'dz'), ('index', 'x'), ('index', 'y'), ('index', 'z')] with shape (4096,)
         [('all', 'particle_ones'), ('index', 'dx'), ('index', 'dy'), ('index', 'dz'), ('index', 'x'), ('index', 'y'), ('index', 'z')] with shape (10,)
```
```suggestion
                these_fields = [
                    field for field, s in field_shapes.items() if s == shape
                ]
```